### ISSUE TITLE
Feature Request: API should return cluster name

### ISSUE THREAD

**Issue by tback** [kubernetes/kubernetes#44954]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):
api cluster name, api get cluster name, cluster name programatically 

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

I want to get the name of the current kubernetes cluster from the api so I can make it available in helm. My usecase is, that I want to annotate automatic backups with the clustername so I don't get name conflicts.

**Comment by liggitt** [comment]
what cluster name are you referring to?

**Comment by tback** [comment]
The one that is shown in `gcloud container clusters list`. Maybe this name isn't currently known inside the cluster, but it'd great if it was.

**Comment by liggitt** [comment]
> Maybe this name isn't currently known inside the cluster, but it'd great if it was.

It is not.

**Comment by tback** [comment]
Drats. Did that happen by choice or by accident? Is there a different name or unique identifier available for a cluster? Is there any chance this'll get implemented?

**Comment by zytek** [comment]
In our use case (clusters in AWS managed by kops) you could 'guess' the cluster name from DNS endpoint (like api.NAME.kubernetes.example.com)

**Comment by liggitt** [comment]
> Is there a different name or unique identifier available for a cluster?

Not one that is provided to, known, or used by the kubernetes processes that make up the cluster, no.

**Comment by kargakis** [comment]
@kubernetes/sig-federation-misc

**Comment by quinton-hoole** [comment]
Kubernetes Cluster Federation has the concept of a cluster name for each cluster which is a member of a Federation.  But it is only known to the federation, not to each cluster. And there is in fact a cluster-id field in the ObjectMeta of every Kubernetes object.    But it is not populated by default.  It is there in order that the Federation control plane can populate it to distinguish API objects originating from different clusters.

It does seem like it would be useful for each cluster to know it's name, but it's not clear to me how this should be exposed, where and how it should be defined, and whether it needs to be unique for all clusters (and if so, how to ensure uniqueness).

**Comment by jfoy** [comment]
@quinton-hoole To further confuse things, kube-controller-manager accepts a `--cluster-name` argument which is used in various places by cloudproviders. 

A formal cluster name as part of the API would be very useful to us, too. In the absence of this property, we're left with various templating hacks at deployment time.  Two use cases that come to mind immediately are:
1. A configuration backup daemon, which needs to push cluster-specific artifacts to some external store.
2. Log and metric collectors, which want to annotate outbound streams with a cluster identifier. 

To my mind, cluster name isn't strictly a federation issue, but a core API design question. Has this idea already been considered and rejected? I haven't found much discussion around it.

**Comment by ereOn** [comment]
I currently manage to get the cluster name by querying the following URL:

```
curl http://metadata/computeMetadata/v1/instance/attributes/cluster-name -H "Metadata-Flavor: Google"
```

Is this not supposed to work ?

**Comment by magescher** [comment]
Using the Python API, I get the current cluster name as follows:
```python
from kubernetes import config

cluster_config = config.kube_config.list_kube_config_contexts()
print cluster_config
([{'name': 'minikube', 'context': {'cluster': 'minikube', 'user': 'minikube'}}], {'name': 'minikube', 'context': {'cluster': 'minikube', 'user': 'minikube'}})

cluster_name = cluster_config[1]['context']['cluster']
print cluster_name
minikube

```

**Comment by liggitt** [comment]
> Using the Python API, I get the current cluster name as follows

That's just the name of the cluster stanza in your local kubeconfig file, and doesn't have any bearing on the server's view of the world

**Comment by geekofalltrades** [comment]
If clusters don't have names that they know about, then how does the `kubernetes.io/cluster/<cluster-name>=owned` tag used by the AWS cloud provider work?

**Comment by irfanurrehman** [comment]
This issue was labelled only for sig/multicluster and is thus moved over to kubernetes/federation#132.
If this does not seem to be right, please reopen this and notify us @kubernetes/sig-multicluster-misc.
/close

**Comment by patrickf55places** [comment]
@irfanurrehman @kubernetes/sig-multicluster-misc this issue is not restricted to clusters using federation. The controller manager has the `--cluster-name` CLI option and at least the AWS cloudconfig file uses a cluster identifier (for tracking cloud resources owned by the cluster). There are many applications that can make use of the identifier (e.g. logging).

In my own use case, I am using [external-dns](https://github.com/kubernetes-incubator/external-dns) which requires an identifier to determine which DNS records are "owned" by the cluster to avoid conflicts with other clusters that use external-dns (e.g. development and production clusters).

**Comment by whereisaaron** [comment]
@irfanurrehman this is an issue for anyone running more that one cluster, it is not specifically a federation issue. I have the same use case as @tback. The ability to tag resources and back-ups with a unique/stable identifier that tracks it back to the cluster.

**Comment by irfanurrehman** [comment]
@whereisaaron @patrickf55places [cluster-registry](https://github.com/kubernetes/cluster-registry) was incubated exactly to store and manage information about the k8s clusters in multi-cluster scenario, not limited to federation. Please check once if that solves your use cases, else:
1 - You can please post your queries at slack either in sig-multicluster (if you see relevance) or kubernetes-dev
2 - Open a new issue at [federation-v2](https://github.com/kubernetes-sigs/federation-v2) with your use case detail (again if you see relevance) or reopen or request to reopen this one here.

**Comment by whereisaaron** [comment]
Thanks @irfanurrehman! I took a look at cluster-registry. I can see how it would be a useful and important component for federation, providing a top-down collection of clusters. However, I couldn't spot any bottom-up identity provided to clusters? If you are a running in a cluster, how do you know what your cluster's identity is within the cluster-registry?

The registry assigns a name based on cluster endpoint(s). If you are in a cluster, the only mechanism to identify yourself I could see, would be if you can identify your external endpoint(s) - which you don't normally need to know - and you also have access to the cluster-registry - then you can enumerate the clusters looking for your endpoint and thus reverse-look-up the name assigned to you. Or is there a better mechanism that I missed?

I don't think cluster-registry addresses e.g. @tback's or my use case of being ability to get a unique cluster identity from the within the cluster. I don't think federation is directly relevant. While federation needs to name/identify clusters, its use case is top-down identity, whereas these use cases are local to each isolated cluster.

Apologies if I have misread how cluster-registry works.

**Comment by irfanurrehman** [comment]
@whereisaaron you understood cluster-registry correctly. What I know is that a [field in the common  object metadata](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L251) was reserved for the info you are looking for, but so far never actually used. I believe there must have been questions to be answered and the api machinery folks could not come to a workable conclusion, or probably no real compelling use cases did show up. 

You might need to post your use case in the slack channel _kubernetes-dev_ or the [developer group](https://groups.google.com/forum/#!topic/kubernetes-dev).

**Comment by ntolia** [comment]
Somewhat related, we too are looking for cluster ID that would be stable for its lifetime (also saw this related but closed issue - https://github.com/kubernetes/kubernetes/issues/2292) but understand that this isn't possible today. As a workaround, I did want to ask whether it is possible for the UID for either the `default` or `kube-system` namespace to ever change. IIUC, those namespaces cannot be deleted and I don't think editing the namespace would change the UID.

**Comment by danielmotaleite** [comment]
@irfanurrehman  There is something wrong here, we are told that the kubernetes cluster do not have the cluster info, yet my aws load balancers have a tag named `KubernetesCluster` with the cluster name... that was created by kubernetes!

so the info is there... so how can we get that?

**Comment by liggitt** [comment]
> @irfanurrehman There is something wrong here, we are told that the kubernetes cluster do not have the cluster info, yet my aws load balancers have a tag named `KubernetesCluster` with the cluster name... that was created by kubernetes!
> 
> so the info is there... so how can we get that?

That is aws-specific, and is provided in the aws cloud provider config file. It is not inherent to the kubernetes cluster.

**Comment by whereisaaron** [comment]
@danielmotaleite sadly that name is known only and used by to the AWS cloud-provider code (which creates the load balancers), so a cloud-specific plug-in, not k8s itself.

**Comment by irfanurrehman** [comment]
For the record, there is some traction on [a proposal](https://docs.google.com/document/d/1F__vEKeI41P7PPUCMM9PVPYY34pyrvQI5rbTJVnS5c4) in SIG Multicluster, which addresses this concern.

**Comment by odinsy** [comment]
There is one way with [yq](https://github.com/mikefarah/yq/) itlity
```
kubectl -n kube-system get configmap kubeadm-config -o jsonpath={.data.ClusterConfiguration} | yq r - clusterName
```

**Comment by dr4Ke** [comment]
Since kubernetes/federation is retired, and there is no reason that even a standalone cluster should not know its name, can this issue be reopened?

**Comment by collimarco** [comment]
It would be extremely useful to have a cluster ID that can be returned by the API. For example an immutable uuid could be created together with the cluster.

**Comment by mukundmckinsey** [comment]
> Using the Python API, I get the current cluster name as follows:
> 
> ```python
> from kubernetes import config
> 
> cluster_config = config.kube_config.list_kube_config_contexts()
> print cluster_config
> ([{'name': 'minikube', 'context': {'cluster': 'minikube', 'user': 'minikube'}}], {'name': 'minikube', 'context': {'cluster': 'minikube', 'user': 'minikube'}})
> 
> cluster_name = cluster_config[1]['context']['cluster']
> print cluster_name
> minikube
> ```

this does not work with incluster_kubeconfig.

**Comment by arielb135** [comment]
> > Using the Python API, I get the current cluster name as follows:
> > ```python
> > from kubernetes import config
> > 
> > cluster_config = config.kube_config.list_kube_config_contexts()
> > print cluster_config
> > ([{'name': 'minikube', 'context': {'cluster': 'minikube', 'user': 'minikube'}}], {'name': 'minikube', 'context': {'cluster': 'minikube', 'user': 'minikube'}})
> > 
> > cluster_name = cluster_config[1]['context']['cluster']
> > print cluster_name
> > minikube
> > ```
> 
> this does not work with incluster_kubeconfig.

Nevertheless, it lists the name of the cluster you defined **locally** in the kubeconfig file, this is not really the cluster name.

**Comment by magescher** [comment]
> 
> Nevertheless, it lists the name of the cluster you defined **locally** in the kubeconfig file, this is not really the cluster name.

@arielb135 Yes, this is a very old thread (4 years old) and I’ve since found better/more stable means.

**Comment by arielb135** [comment]
> > Nevertheless, it lists the name of the cluster you defined **locally** in the kubeconfig file, this is not really the cluster name.
> 
> @arielb135 Yes, this is a very old thread (4 years old) and I’ve since found better/more stable means.

Do share? I'm tackling the same issue and I prefer not doing per vendor approach, or some random guid

**Comment by Varun-garg** [comment]
what should be the source of truth here - Kube config, cloud vendor, or cluster?

**Comment by fzyzcjy** [comment]
Is there any updates? This is very helpful and is not yet addressed (but is closed :( )

**Comment by ait-maksymp** [comment]
This is pathetic. People are selling this hipster technology as if it was the only answer to any computing need in the universe, but still this tool can't tell me its own name!

**Comment by imperialguy** [comment]
This is bad. A cluster that can't find it's own id or name from within one of its pods? This needs to be a mandatory feature. It's almost 2023, still not solution. Any pod/container running inside a k8s cluster must be able to find out the id and name of the cluster that created it.

**Comment by EronWright** [comment]
To make progress, may I suggest that a proposal be raised in the "[Cluster API](https://github.com/kubernetes-sigs/cluster-api)" project, e.g. to make the cluster name be available at runtime.

**Comment by mikeshng** [comment]
Thread in the #sig-multicluster Slack:
https://kubernetes.slack.com/archives/C09R1PJR3/p1701167487555119

TLDR: worth taking a look at the following:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/2149-clusterid#clusterproperty-crd
https://github.com/kubernetes-sigs/about-api

**Comment by caozhuozi** [comment]
any updates in 2024?

**Comment by dray92** [comment]
how can we get this issue prioritized or looked at? it feels weird that workloads need some elevated privileges (eg: inst metadata access in eks), or undergo massively fragile strategies to discover cluster name by dissecting labels etc from nodes to obtain this info.
